### PR TITLE
Adding Spatter

### PR DESCRIPTION
**Pull request by mahyarsamani**
This PR adds source code for C++ implementation of SpatterGen as well as SpatterKernel. SpatterGen uses a PyBindMethod to add kernels to the backend code. This way the process of processing json files could be offloaded to python. In addition it adds standard library components for SpatterGenCore and SpatterGen. These two components follow the same structure as AbstractCore and AbstractProcessor. In addition spatter_kernel.py adds a definition for SpatterKernel in python to make adding kernels to C++ easier. Also it adds utility functions for parsing dictionaries read from json as well as partitioning traces for multicore setups.

### CODE REVIEW

**Comment by powerjg** [comment]
I know this is a draft, but a few quick comments:
1. You need to add copyright blocks
2. The names should be something more descriptive than "model"
3. The commit message and the PR description need to have details

File: src/base/stats/units.hh
Comment (mahyarsamani) [inline src/base/stats/units.hh]: @powerjg If I want to express latency stats in seconds, due to precision limitations they will all be zero. I added this new unit to express the latencies in nanoseconds. I am not really sure if this is what the right way to do this is. Let me know if I have to change this to Tick.

Comment (powerjg) [reply src/base/stats/units.hh]: Yeah, I'm against this. Just use `Ticks` if you need high precision.

File: src/cpu/testers/spatter_gen/SpatterGen.py
Comment (powerjg) [inline src/cpu/testers/spatter_gen/SpatterGen.py]: Can you add a reference to the spatter paper for this class? Call `add_citation(SpatterGen, <bibtex text>)` at the end of this file.

File: src/cpu/testers/spatter_gen/SpatterGen.py
Comment (powerjg) [inline src/cpu/testers/spatter_gen/SpatterGen.py]: Should this be a `Param.Cycles` if it's cycles?

File: src/cpu/testers/spatter_gen/SpatterGen.py
Comment (powerjg) [inline src/cpu/testers/spatter_gen/SpatterGen.py]:
```suggestion
    request_gen_rate = Param.Int("Number of requests generate per cycle.")
```
Bandwidth isn't exactly the best term here.

File: src/cpu/testers/spatter_gen/SpatterGen.py
Comment (powerjg) [inline src/cpu/testers/spatter_gen/SpatterGen.py]:
```suggestion
    request_buffer_entries = Param.Int("Size of the request buffer.")
```
To me, `size` implies it is in `Bytes`

File: src/cpu/testers/spatter_gen/SpatterGen.py
Comment (powerjg) [inline src/cpu/testers/spatter_gen/SpatterGen.py]: I don't understand how this is different from `request_gen_bandwidth`. The name and/or help text needs to be updated to make it more clear.

File: src/cpu/testers/spatter_gen/enums.cc
Comment (powerjg) [inline src/cpu/testers/spatter_gen/enums.cc]: You shouldn't need this file at all. See the code here for how it works: https://github.com/gem5/gem5/blob/stable/src/base/stl_helpers/ostream_helpers.hh#L81

File: src/cpu/testers/spatter_gen/spatter_gen.hh
Comment (powerjg) [inline src/cpu/testers/spatter_gen/spatter_gen.hh]: Can you use a less generic name, like "SpatterGenEvent"? I think I see how you want this to be generic, but I don't think it makes sense to make it generic here and now 🙂

File: src/cpu/testers/spatter_gen/spatter_gen.hh
Comment (powerjg) [inline src/cpu/testers/spatter_gen/spatter_gen.hh]:
```suggestion
        SpatterGen::State _state;
```

I would make this private to the `SpatterGen` class.

File: src/cpu/testers/spatter_gen/enums.hh
Comment (powerjg) [inline src/cpu/testers/spatter_gen/enums.hh]: I would make this private to the `SpatterGen` class

File: src/cpu/testers/spatter_gen/spatter_gen.hh
Comment (powerjg) [inline src/cpu/testers/spatter_gen/spatter_gen.hh]: I don't understand this comment

File: src/cpu/testers/spatter_gen/spatter_gen.hh
Comment (powerjg) [inline src/cpu/testers/spatter_gen/spatter_gen.hh:213]: Why is this a vector?

File: src/cpu/testers/spatter_gen/spatter_gen.hh
Comment (powerjg) [inline src/cpu/testers/spatter_gen/spatter_gen.hh]: I'm concerned these are queues. Do we have guarantees on ordering?

It also seems like there are pairs of `PacketPtr` and `Tick`. Why not use one datastructure?

File: src/python/gem5/components/processors/spatter_generator.py
Comment (powerjg) [inline src/python/gem5/components/processors/spatter_generator.py]: It would be good to add a comment here about why you don't need an `else` statement.

File: src/python/gem5/components/processors/spatter_generator_core.py
Comment (powerjg) [inline src/python/gem5/components/processors/spatter_generator_core.py]: Probably should drop this print for upstream.

File: src/python/gem5/components/processors/spatter_kernel.py
Comment (powerjg) [inline src/python/gem5/components/processors/spatter_kernel.py]: This class should really come with some help comments explaining how it's used. Pointers to spatter documentation is great.

File: src/python/gem5/components/processors/spatter_generator.py
Comment (powerjg) [inline src/python/gem5/components/processors/spatter_generator.py]: I would put these files in a subdirectory. You can expose them in the `processors` module via `__init__.py` if you want, but this directory is getting out of hand.

File: src/cpu/testers/spatter_gen/spatter_gen.cc
Comment (powerjg) [inline src/cpu/testers/spatter_gen/spatter_gen.cc]: Is this how the other generators work? I thought if they had nothing to do, they would just not schedule any events and you would get a `MaxTick` exit. I would like all generators to have the same behavior.

File: src/cpu/testers/spatter_gen/spatter_gen.cc
Comment (powerjg) [inline src/cpu/testers/spatter_gen/spatter_gen.cc]:
```suggestion
void
SpatterGen::SpatterGenPort::sendPacket(PacketPtr pkt)
{
    panic_if(blocked(), "Should never try to send if port is blocked.");
```

File: src/cpu/testers/spatter_gen/spatter_gen.cc
Comment (powerjg) [inline src/cpu/testers/spatter_gen/spatter_gen.cc]: Is this a future todo?

File: src/cpu/testers/spatter_gen/spatter_gen.cc
Comment (powerjg) [inline src/cpu/testers/spatter_gen/spatter_gen.cc]: Would it make sense to put a panic/assert here?

File: src/cpu/testers/spatter_gen/spatter_gen.cc
Comment (powerjg) [inline src/cpu/testers/spatter_gen/spatter_gen.cc]: Why is this not done in `checkSimDone`?

File: src/cpu/testers/spatter_gen/spatter_gen.cc
Comment (powerjg) [inline src/cpu/testers/spatter_gen/spatter_gen.cc]: I would split these two `if`/`else if` clauses into their own functions. This function is too big.

**Review by powerjg — COMMENTED** [review]
Overall, this looks great. I didn't check the logic carefully in `spatter_gen.cc`, but I trust you've tested this and compared with upstream spatter.

Most of my comments are style. The only "big" thing that I would like to see changed is moving the state information into the `SpatterGen` class to make it private. 

One thing to keep in mind when pushing code upstream is that you should try to make things as self-contained as possible. Don't expose things publicly unless absolutely required.

Finally, there are some places where you need to add more documentation. Think about how someone would discover how to use this if they don't know anything about Spatter and they are not involved in our project.

File: src/cpu/testers/spatter_gen/spatter_gen.cc
Comment (mahyarsamani) [reply src/cpu/testers/spatter_gen/spatter_gen.cc]: checkSimDone is for overall exit. This exit is per kernel. We're not completely done here. I think it is useful to try dump stats per kernel.

File: src/cpu/testers/spatter_gen/spatter_gen.cc
Comment (mahyarsamani) [reply src/cpu/testers/spatter_gen/spatter_gen.cc]: I agree. I would argue for not changing it right now. I will update this after we have a more definitive picture of what we want the SpatterKernel class to look like.

File: src/cpu/testers/spatter_gen/spatter_gen.cc
Comment (mahyarsamani) [reply src/cpu/testers/spatter_gen/spatter_gen.cc]: I will put a panic telling user to make fpRegFileSize bigger than request_buffer_entries.

File: src/cpu/testers/spatter_gen/spatter_gen.hh
Comment (mahyarsamani) [reply src/cpu/testers/spatter_gen/spatter_gen.hh]: Answer to second question is, it's a preference.
About your first question, do you mean that responses do not arrive at the same order requests were issued? If so, I can ensure that all responses before an expected response have arrived before another access is made based on the resp.

File: src/cpu/testers/spatter_gen/spatter_gen.hh
Comment (mahyarsamani) [reply src/cpu/testers/spatter_gen/spatter_gen.hh:213]: I need to access it with index.

**Comment by mahyarsamani** [comment]
@powerjg @giactra 
Still working on addressing some comments. Will fix by tomorrow.

File: src/cpu/testers/spatter_gen/spatter_gen.cc
Comment (mahyarsamani) [reply src/cpu/testers/spatter_gen/spatter_gen.cc]: Yes

File: src/python/gem5/components/processors/spatter_gen/__init__.py
Comment (powerjg) [inline src/python/gem5/components/processors/spatter_gen/__init__.py:28]: Copyright...

File: src/cpu/testers/spatter_gen/spatter_gen.cc
Comment (powerjg) [inline src/cpu/testers/spatter_gen/spatter_gen.cc]: You don't need these for loops. You can set the value to use when calling `resize`. See https://en.cppreference.com/w/cpp/container/vector/resize

File: src/cpu/testers/spatter_gen/spatter_gen.cc
Comment (powerjg) [reply src/cpu/testers/spatter_gen/spatter_gen.cc]: What do you think about this?

File: src/cpu/testers/spatter_gen/spatter_gen.cc
Comment (powerjg) [reply src/cpu/testers/spatter_gen/spatter_gen.cc]: I would argue that if we don't know what we want the kernel to look like, then it's not ready to be merged upstream. We should go ahead and do it the right way the first time. Once something is merged, the activation energy to change it is very high.

File: src/cpu/testers/spatter_gen/spatter_gen.hh
Comment (powerjg) [reply src/cpu/testers/spatter_gen/spatter_gen.hh]: Don't forget about this.

File: src/python/gem5/components/processors/spatter_gen/spatter_kernel.py
Comment (powerjg) [inline src/python/gem5/components/processors/spatter_gen/spatter_kernel.py]: Can you point to the repo instead? I don't trust that bitly will be around in 20 years.

File: src/cpu/testers/spatter_gen/spatter_gen.cc
Comment (mahyarsamani) [reply src/cpu/testers/spatter_gen/spatter_gen.cc]: You are right, this behavior is different. However, I remember having much trouble with using bandwidth stats with that behavior.

File: src/cpu/testers/spatter_gen/spatter_gen.cc
Comment (powerjg) [inline src/cpu/testers/spatter_gen/spatter_gen.cc]: `fatal`? If this is a user configuration error it should be fatal. Panic is for bugs (like an assert)

File: src/cpu/testers/spatter_gen/spatter_gen.cc
Comment (powerjg) [inline src/cpu/testers/spatter_gen/spatter_gen.cc:110]: I would add an `assert(blockedPacket);` here

File: src/cpu/testers/spatter_gen/spatter_gen.hh
Comment (powerjg) [inline src/cpu/testers/spatter_gen/spatter_gen.hh]: 1. Where are these defined?
2. Why are they not in the `SpatterGen` class? Do you have any uses `TimedPacket` outside of the class members?

File: src/cpu/testers/spatter_gen/spatter_gen.hh
Comment (powerjg) [inline src/cpu/testers/spatter_gen/spatter_gen.hh]: Seems like this could just be a boolean `isPending` (or `isWaiting` or something like that)

File: src/cpu/testers/spatter_gen/spatter_gen.hh
Comment (powerjg) [inline src/cpu/testers/spatter_gen/spatter_gen.hh]: What is the difference between `pending()` and `sleep()`?

**Review by powerjg — COMMENTED** [review]
A few more little things

File: src/cpu/testers/spatter_gen/spatter_gen.hh
Comment (mahyarsamani) [reply src/cpu/testers/spatter_gen/spatter_gen.hh]: pending is a getter and sleep is a setter.

File: src/cpu/testers/spatter_gen/spatter_gen.hh
Comment (mahyarsamani) [reply src/cpu/testers/spatter_gen/spatter_gen.hh]: I like to keep it an enum, I know right now the state is a boolean like value. In future I want to extend this so that it can encode more states, e.g. ASLEEP_ON_INPUT, ASLEEP_ON_OUTPUT, ...

File: src/cpu/testers/spatter_gen/spatter_gen.hh
Comment (mahyarsamani) [reply src/cpu/testers/spatter_gen/spatter_gen.hh]: They are defined in spatter_gen.cc. It's a matter of preference for me to not define it in the SpatterGen class. I can move it inside if there is a right/wrong.

File: src/cpu/testers/spatter_gen/spatter_gen.cc
Comment (mahyarsamani) [reply src/cpu/testers/spatter_gen/spatter_gen.cc]: Will do

File: src/cpu/testers/spatter_gen/spatter_gen.cc
Comment (mahyarsamani) [reply src/cpu/testers/spatter_gen/spatter_gen.cc:110]: Will do

File: src/cpu/testers/spatter_gen/spatter_gen.hh
Comment (powerjg) [reply src/cpu/testers/spatter_gen/spatter_gen.hh]: They are polluting the gem5 namespace is the problem. I would put them in the class until they are used outside of it.

Comment (mahyarsamani) [reply src/cpu/testers/spatter_gen/spatter_gen.hh]: I see, will do.

**Review by powerjg — DISMISSED** [review]
Looks good to me. Thanks for the updates!

**Comment by mahyarsamani** [comment]
@giactra This is ready for review.

**Review by powerjg — CHANGES_REQUESTED** [review]
There's a merge/rebase issue here.

**Comment by mahyarsamani** [comment]
> There's a merge/rebase issue here.

I think it should be fixed now.

**Review by powerjg — DISMISSED** [review]

**Comment by powerjg** [comment]
@giactra let me know if you want us to hold off merging until you have time to review.

File: src/cpu/testers/spatter_gen/spatter_gen.hh
Comment (giactra) [inline src/cpu/testers/spatter_gen/spatter_gen.hh:78]: I appreciate there is a reference to the publication in the SpatterGen.py file, but I believe it would be beneficial for a user to have a brief description of Spatter classes and what they are actually doing, so that a user is not "forced" to read the paper, especially if the paper requires a subscription

File: src/cpu/testers/spatter_gen/spatter_gen.hh
Comment (giactra) [inline src/cpu/testers/spatter_gen/spatter_gen.hh]: Is there a reason to make them available like this in the gem5 namespace? Shall we restrict the visibility to the Spatter class?

File: src/cpu/testers/spatter_gen/spatter_gen.hh
Comment (giactra) [inline src/cpu/testers/spatter_gen/spatter_gen.hh]: I would use scoped enum here

File: src/cpu/testers/spatter_gen/spatter_gen.hh
Comment (giactra) [inline src/cpu/testers/spatter_gen/spatter_gen.hh]: Nitpick: this could be marked as const

File: src/cpu/testers/spatter_gen/spatter_gen.hh
Comment (giactra) [inline src/cpu/testers/spatter_gen/spatter_gen.hh]: Ditto

File: src/cpu/testers/spatter_gen/utility_structs.hh
Comment (giactra) [inline src/cpu/testers/spatter_gen/utility_structs.hh]: Ditto

File: src/cpu/testers/spatter_gen/utility_structs.hh
Comment (giactra) [inline src/cpu/testers/spatter_gen/utility_structs.hh]: I would move this within the SpatterAccess class

File: src/cpu/testers/spatter_gen/utility_structs.hh
Comment (giactra) [inline src/cpu/testers/spatter_gen/utility_structs.hh]: This is not matching the header file

File: src/cpu/testers/spatter_gen/utility_structs.hh
Comment (giactra) [inline src/cpu/testers/spatter_gen/utility_structs.hh]: Could be marked as const

File: src/cpu/testers/spatter_gen/utility_structs.hh
Comment (giactra) [inline src/cpu/testers/spatter_gen/utility_structs.hh]: I would pass by const reference here as you are making a copy

File: src/cpu/testers/spatter_gen/utility_structs.hh
Comment (giactra) [inline src/cpu/testers/spatter_gen/utility_structs.hh]: could be const

File: src/cpu/testers/spatter_gen/utility_structs.hh
Comment (giactra) [inline src/cpu/testers/spatter_gen/utility_structs.hh]: could be const

File: src/cpu/testers/spatter_gen/utility_structs.hh
Comment (giactra) [inline src/cpu/testers/spatter_gen/utility_structs.hh]: Could be const

File: src/cpu/testers/spatter_gen/utility_structs.hh
Comment (giactra) [inline src/cpu/testers/spatter_gen/utility_structs.hh]: Could be const

**Review by giactra — DISMISSED** [review]
Thanks @powerjg, I am not going to do a full functionality check assuming it has been tested and works properly already. I only wanted to make sure the code was properly documented (that's the main comment I have if you see below). Also some minor comments on the code from a SW engineering point of view

File: src/cpu/testers/spatter_gen/spatter_gen.hh
Comment (mahyarsamani) [reply src/cpu/testers/spatter_gen/spatter_gen.hh:78]: Would the documentation in spatter_kernel.py suffice? I am now looking at writing it up and wonder if I should extend what is already there.

**Comment by BobbyRBruce** [comment]
@powerjg : Can you sign off on this so we can get it in for the v24.0 staging?

**Review by BobbyRBruce — APPROVED** [review]

**Pull request merged**
